### ISSUE TITLE
`np` has some good ideas. Let's "borrow" them.

### ISSUE THREAD

**Issue by jamestalmage** [bjoerge/cut-release#13]
Check out [`np`](https://github.com/sindresorhus/np), it does some stuff we don't.
- Checks for unpulled changes on the git remote.
- removes and reinstalls dependencies
- uses `trash-cli` to remove files (moves to trash bin, instead of gone forever).

// @sindresorhus

**Comment by alexkwolfe** [comment]
Some other ideas from our fork of cut-release:
- Ensure that the [local repo is tracking a remote](https://github.com/activeprospect/cut-release/blob/master/bin/cut-release.js#L250-L299)
- Ensure that there are [no uncommitted changes locally](https://github.com/activeprospect/cut-release/blob/master/bin/cut-release.js#L472-L482)
- Handle the situation where [git has already been tagged](https://github.com/activeprospect/cut-release/blob/master/bin/cut-release.js#L413-L445)

We also [ensure that local isn't ahead of the branch being tracked](https://github.com/activeprospect/cut-release/blob/master/bin/cut-release.js#L447-L470).

We've been using these for a while and I've been thinking about doing PRs for them.

**Comment by jamestalmage** [comment]
My thoughts.
- Ensure local repo is tracking a remote. :+1: 
- Ensure that there are no uncommitted changes locally. `npm version` (which we call) - already does that. What value is there in having `cut-release` do the same?
- Handle situation where git has already been tagged: :+1: 
- Ensure that local is not ahead of tracked branch :-1: - Why? If I am ahead of the remote, I am not going to have a push conflict. Seems an overly opinionated choice.

What about ensuring the remote tracked branch is `master`? Are people cutting releases on any branch except master? Seems odd if they are. I could definitely see myself screwing up and calling while on a PR branch though.

**Comment by alexkwolfe** [comment]
> Ensure that local is not ahead of tracked branch :-1: - Why? If I am ahead of the remote, I am not going to have a push conflict. Seems an overly opinionated choice.

We do PRs and run tests in Travis. This test helps ensure that a developer isn't publishing commits that have never been seen by the rest of the development group and/or have never been run in Travis.

**Comment by alexkwolfe** [comment]
> Ensure that there are no uncommitted changes locally. npm version (which we call) - already does that. What value is there in having cut-release do the same?

We were relying on the `npm version` failure for a while, as you suggest, but found it somewhat annoying to have it fail for this reason after answering all the cut-release prompts. Our fork does this validation up front before any questions are asked, and bails with a sensible error if the user has uncommitted changes.

**Comment by jamestalmage** [comment]
OK, I buy the second argument (checking for a clean git tree up front). Not sure on the other one.

**Comment by alexkwolfe** [comment]
Yeah. I'm not sure if testing that everything is pushed prior to cutting a new release is of value to other folks. This was an actual problem we had, so I added it. Seems like maybe it's only useful for teams, and even then maybe not for all teams.

**Comment by sindresorhus** [comment]
> We do PRs and run tests in Travis. This test helps ensure that a developer isn't publishing commits that have never been seen by the rest of the development group and/or have never been run in Travis.

I [initially](https://github.com/sindresorhus/np/commit/3b2a7665fc00845f704ede7af2a7563f3ddb1204) had that in `np`, but personally found it annoying. I'm sure it's useful for teams, though. Maybe make it optional.

**Comment by alexkwolfe** [comment]
After looking at this more closely, I realized that our branch doesn't do what I said it does. It only checks that the remote isn't ahead. I guess it wasn't that big of a problem, after all. :)

**Comment by jamestalmage** [comment]
Answering my own question:

> What about ensuring the remote tracked branch is master?

:-1: because: "release branches", and sometimes you want to backport fixes.

If it is not master, then maybe we could add some simple logic to validate what they are doing makes sense, and ask them if it doesn't.

Like, if the branch name looks like semver (i.e. `0.6.x`), then warn if they are trying to do a bump that would take them outside the implied semver range (i.e. anything outside a patch release in this case).

That is probably a lot of work, though. We should focus on the other stuff first.

@alexkwolfe - Can you start submitting PR's from your branch for the features where we have consensus?

**Comment by koresar** [comment]
Hello gentlemen.

Just wanted to say thank you without opening a new issue.

# Thank you

you f*cking rock!
